File: 백준/Gold/10026. 적록색약/적록색약.py
n = int(input())
grid = [list(input().strip()) for _ in range(n)]

visited1 = [[False] * n for _ in range(n)]
visited2 = [[False] * n for _ in range(n)]

count1 = 0
count2 = 0

dx = [-1,1,0,0]
dy = [0,0,-1,1]

from collections import deque
def bfs(x,y,initialcolor,visited,isblind):
    queue = deque()
    visited[x][y] = True
    queue.append((x,y))
    while queue:
        cx,cy = queue.popleft()

        for d in range(4):
            nx = cx + dx[d]
            ny = cy + dy[d]

            if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny]:
                if isblind:
                    if (initialcolor == 'R' and grid[nx][ny] == 'G') or (initialcolor == 'G' and grid[nx][ny] == 'R') or (initialcolor == grid[nx][ny]):
                        visited[nx][ny] = True
                        queue.append((nx,ny))
                else:
                    if initialcolor == grid[nx][ny]:
                        visited[nx][ny] = True
                        queue.append((nx,ny))
    return 1

for i in range(n):
    for j in range(n):
        if not visited1[i][j]:
            count1 += bfs(i,j,grid[i][j],visited1,False)
        if not visited2[i][j]:
            count2 += bfs(i,j,grid[i][j],visited2,True)

print(count1, count2)

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.*;
import java.util.Deque;
import java.util.ArrayDeque;

class Point {
    int x,y;
    Point (int x, int y){
        this.x = x;
        this.y = y;
    }
}

public class Main {
    static int[] dx = {-1,1,0,0};
    static int[] dy = {0,0,-1,1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        String[][] grid = new String[n][n];

        for (int i = 0; i < n; i++){
            String s = br.readLine();
            for (int j = 0; j < n; j ++){
                grid[i][j] = String.valueOf(s.charAt(j));
            }
        }

        boolean[][] visited = new boolean[n][n];
        boolean[][] visited2 = new boolean[n][n];

        int count1 = 0;
        int count2 = 0;

        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j ++) {
                if (!visited[i][j]) {
                    count1 += bfs(i, j, grid, visited, false, n);
                }
                if (!visited2[i][j]){
                    count2 += bfs(i,j,grid, visited2, true, n);
                }
            }
        }

        System.out.println(count1 + " " + count2);
    }

    static int bfs(int i, int j, String[][]grid, boolean[][] visited, boolean isBlind, int n){
        Deque<Point> queue = new ArrayDeque<>();
        queue.offerLast(new Point(i,j));
        visited[i][j] = true;
        String color = grid[i][j];

        while (!queue.isEmpty()){
            Point p = queue.pollFirst();
            int x = p.x;
            int y = p.y;

            for (int d = 0; d < 4; d ++){
                int nx = x + dx[d];
                int ny = y + dy[d];
                // 범위내, 미방문, 컬러 구분
                 if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny]) {
                    if (!isBlind) {
                        // 정상인
                        if (grid[nx][ny].equals(color)) {
                            visited[nx][ny] = true;
                            queue.offerLast(new Point(nx, ny));
                        }
                    } else {
                        // 적록색약
                        if (color.equals(grid[nx][ny]) ||
                            (color.equals("R") && grid[nx][ny].equals("G")) ||
                            (color.equals("G") && grid[nx][ny].equals("R"))) {
                            visited[nx][ny] = true;
                            queue.offerLast(new Point(nx, ny));
                        }
                    }
                }
            }
        }
        return 1;
    }


}